### ISSUE TITLE
Do we have full recursive validity for references

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#412]
This is one of the successors to https://github.com/rust-lang/unsafe-code-guidelines/issues/77. The question is: for a reference to be valid, do we require that all the data it points to is equally valid, fully recursively?

I think the answer is "no", and my impression is most people agree, so let me collect some arguments here (and I will open some other issues that only make sense discussing after the answer here is "no"):
- Full recursive validity is super fragile: to reason about why the reference you are creating does not cause UB, you have to reason about all code that might mutate any memory transitively reachable (via references).
- The benefit is questionable: having deeply nested references cause UB is very unlikely to ever benefit compiler optimizations.
- Allowing mutable references to uninit data (between consenting APIs) is (unfortunately) how the `io::Read` trait often works, and while better solutions are being developed, there's a long tail of copies of this API and a lot of old code to port, that we should have good reasons to consider UB.
- For shared references with interior mutability specifically, even determining whether an `&Mutex<bool>` is valid would cause a conceptual data race with another thread holding the lock and mutating that `bool`.
- Some of the arguments in https://github.com/rust-lang/unsafe-code-guidelines/issues/346 also apply, though some are more specific and a lot of the discussion focuses not on "full recursive validity: yes or know" but various weaker validity variants.

**Comment by digama0** [comment]
I think point 4 means that "full recursive validity" is not actually well-defined. A reference does not necessarily give access to the full pointed-to allocation, if parts of that allocation are currently loaned out, and I think one aspect of the memory model that we should all be able to agree on is that a evaluating a validity assertion should not require "reads" which are not legal by the memory model itself, because this leads to data-race-like issues which I'm not sure can be resolved in the presence of weak memory.

However, not all validity properties are incompatible with concurrent reads, because not everything is under an `UnsafeCell`. I would like us to separate the extremes of "full recursive validity" and "validity of references cannot depend on memory". I would say that "full recursive validity" is not a coherent position as currently understood, but there is still a possibility for a middle ground where some memory is involved in the validity of references, possibly even multiple levels of pointer indirection, as long as we do not need to look at memory which is currently loaned out or memory behind an `UnsafeCell`.

**Comment by RalfJung** [comment]
> TL;DR: The answer to the title question should be "no", but the follow up questions might not be so simple.

Yes. The idea was to have this issue for documenting that consensus, so that we can have separate issues for the follow-up questions. That is my attempt to disentangle the various follow-up questions that arise here.

Let's collect follow-up questions we have issues for:
- https://github.com/rust-lang/unsafe-code-guidelines/issues/346
- https://github.com/rust-lang/unsafe-code-guidelines/issues/412
- https://github.com/rust-lang/unsafe-code-guidelines/issues/236
- as well as the [issues tagged `A-dereferenceable`](https://github.com/rust-lang/unsafe-code-guidelines/labels/A-dereferenceable)

Please open a new issue if you think there is a separate proposal/discussion to be had.

**Comment by celinval** [comment]
Is this issue proposing to change the [UB defined in the reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)?

> A reference or Box<T> that is [dangling](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers), misaligned, or **points to an invalid value**.

**Comment by chorman0773** [comment]
It would still be an unsafe value, but yes, it would not be undefined behaviour to produce the reference.

**Comment by RalfJung** [comment]
Yes, many of the issues in this repo are about finding out what UB we really want. The reference list is preliminary, as indicated by the big warning there.

**Comment by RalfJung** [comment]
Turns out at least recursive validity if `Box` is violated all the time, since the MIR we generate for `Box::new` first creates a fresh initialized box and then stores a `T` in there -- so we have a `Box<T>` value pointing to uninitialized memory for a short time.

For non-`Box`, the first issue I am hitting is [this code](https://github.com/rust-lang/rust/blob/0a42a540c603846aa22f29f378a61a64c9d4383e/library/core/tests/io/borrowed_buf.rs#L137-L142) which blatantly tells `BorrowedBuf` that there are 8 initialized bytes when that's not actually true, and subsequently calls `init_ref`, creating a slice to uninit memory.

And then, the first issue that's not just in a test: [this code here](https://github.com/rust-lang/rust/blob/cf87203f483751e0f6daa74b9d129c04feb48355/library/core/src/slice/mod.rs#L4469-L4472) is unsound with recursive reference validity. `*(*arr_ptr).get_unchecked_mut(i)` will turn `arr_ptr: *mut [&mut T; N]` into a reference to pass it to `get_unchecked_mut`, but the array has not been fully initialized yet.

**Comment by RalfJung** [comment]
Full recursive validity is also fundamentally at odds with `#[may_dangle]`. This code is UB under full recursive validity:
```rust
#![feature(dropck_eyepatch)]

struct Foo<'a>(&'a u8);

unsafe impl<#[may_dangle] 'a> Drop for Foo<'a> {
    fn drop(&mut self) { }
}

fn main() {
    let foo;
    {
        let x = 0;
        foo = Foo(&x);
    }
}
```
(Example due to @JakobDegen, from a different discussion )

The `self` argument in `drop` will be `&mut Newtype(&'a u8)` , where `'a` is dead. So full recursive validity implies that
-  either `#[may_dangle]` cannot be used on lifetimes/types that "really occur" in the type (i.e., it can only be used like in `Vec` where the relevant `T` is behind a raw pointer and inside `PhantomData`, but not directly safely reachable)
- or `#[may_dangle]` affects the validity requirements of the `drop` function it is attached to -- which is a ticking bomb since any function called by `drop` does not inherit this special treatment

**Comment by workingjubilee** [comment]
> The benefit is questionable: having deeply nested references cause UB is very unlikely to ever benefit compiler optimizations.

That part seems speculative. I have absolutely seen code that looks like
```rust
let iter = vec.iter().filter(pred);
call();
for item in iter.filter(overlapping_pred) {
    // code
}
```

in the wild... usually because the first part looks like calling this:

```rust
impl SomeType<T> {
    fn iter(&self) -> impl Iterator<Item = &T> {
         self.inner_slice.filter(valid_elements)
    }
}
```

And wouldn't that benefit from the optimization that's currently invalid in https://github.com/rust-lang/rust/issues/130853 of a simple `&&u8`?

```rust
let val = **x;
call();
**x == val // optimize to true
```

Obviously this may not hold for all types.

**Comment by RalfJung** [comment]
Aliasing for nested references and recursive validity do not *have to* be entangled, so please keep the former to https://github.com/rust-lang/unsafe-code-guidelines/issues/532.

**Comment by workingjubilee** [comment]
*squints*

Huuuh, okay!

**Comment by RalfJung** [comment]
Closing this in favor of https://github.com/rust-lang/unsafe-code-guidelines/issues/414 since the issues have a confusing amount of overlap and the maximalist position that this issue asks about [doesn't work with current Rust](https://github.com/rust-lang/unsafe-code-guidelines/issues/412#issuecomment-2379009718) isn't getting much support.
